src/main/ml-modules/ext/serviceModel.sjs

```diff
@@ -71,41 +73,924 @@ function getServiceModels(filter) {
+    //avoids the overhead of pulling the full descriptor doc.
+    let uri = fn.head(cts.uris(null, null, modelQuery));
+    if (DEBUG) xdmp.trace("GDS-DEBUG", "getServiceModelIndexEntry(): serviceModelIndex[serviceId]: uri = " + uri);
+    setServiceModelIndexEntry(null, uri, serviceId);
   }
+  if (DEBUG) xdmp.trace("GDS-DEBUG", "getServiceModelIndexEntry(): returning index entry");
+  return serviceModelIndex[serviceId];
 }
 
-function getServiceModel(serviceId) {
-  return getServiceModelInfo(serviceId).model;
+/**
+ * This function wraps a service descriptor doc in an index entry.  If modelDoc is null, it will use
+ * the uri and serviceName params to create a lightweight shell entry.
+ * 
+ * @param modelDoc {Node} the service model document from disk, may be null if uri and service name are provided
+ * @param uri {String} @default [null] The uri of the service model document. Used when creating a lightweight entry.    
+ * @param serviceName {String} @default [null] The service's service name/id. Used when creating a lightweight Entry.
+ */
+function setServiceModelIndexEntry(modelDoc, uri = null, serviceName = null) {
+  if (modelDoc == null && uri != null) {
+    if (DEBUG) xdmp.trace("GDS-DEBUG", "setServiceModelIndexEntry(); creating shallow index entry")
+    serviceModelIndex[serviceName] = {
+      serviceModel:null,
+      serviceDescriptor:null,
+      layerModelIndex:{},
+      uri:uri
+    };
+  } 
+  else {
+    let modelObj = modelDoc.toObject();
+    let serviceId = modelObj.info.name;
+    if (serviceModelIndex[serviceId] != null) {
+      serviceModelIndex[serviceId].serviceModel = modelObj;
+    }
+    else {
+      serviceModelIndex[serviceId] = {
+          serviceModel:modelObj,
+          serviceDescriptor:null,
+          layerModelIndex:{},
+          uri:xdmp.nodeUri(modelDoc)
+        };
+    }
+  }
 }
 
+/**
+ * Returns the layer model for the given service name/id and layer id.
+ * @param serviceName {String} the service name/id
+ * @param layerId {int} the layer number.
+ */
+function getLayerModel(serviceName, layerId) {
+  if (DEBUG) xdmp.trace("GDS-DEBUG", "Starting getLayerModel");
+
+  //const serviceModel = getServiceModel(serviceName);
+  let layerModelIndexEntry = getLayerModelIndexEntry(serviceName, layerId);
+
+  return layerModelIndexEntry.layerModel;
+}
+
+function setFieldDescriptors(layer, serviceModel) {
+  if (layer.gdsMetadata == null || layer.gdsMetadata.columnDefs == null || viewUpdated(layer)) {
+    if (DEBUG) xdmp.trace("GDS-DEBUG", "view(s) updated, regenerating field descriptors");
+    let columnDefs = generateFieldDescriptors(layer, serviceName)
+    if (layer.gdsMetadata == null) layer.gdsMetadata = {};
+    layer.gdsMetadata.viewHashes = getViewHashes(layer);
+    layer.gdsMetadata.columnDefs = columnDefs;
+
+    // xdmp.invokeFunction(function() {
+    //   saveServiceModel(serviceName, serviceModel);
+    // },
+    // {isolation:"different-transaction"}
+    // );
```

Comment:
Why is this commented out? If it can be deleted, remove it or add a comment as to why it is still there.